src/ripple/app/paths/impl/BookStep.cpp

```diff
@@ -156,34 +162,269 @@ class BookStep : public StepImp<TIn, TOut, BookStep<TIn, TOut>>
 
     bool equal (Step const& rhs) const override;
 
+    // Iterate through the offers at the best quality in a book.
+    // Unfunded offers and bad offers are skipped (and returned).
+    // callback is called with the offer SLE, taker pays, taker gets.
+    // If callback returns false, don't process any more offers.
+    // Return the unfunded and bad offers and the number of offers consumed.
+    template <class Callback>
+    std::pair<boost::container::flat_set<uint256>, std::uint32_t>
+    forEachOffer (
+        PaymentSandbox& sb,
+        ApplyView& afView,
+        bool prevStepRedeems,
+        Callback& callback) const;
+
     void consumeOffer (PaymentSandbox& sb,
         TOffer<TIn, TOut>& offer,
         TAmounts<TIn, TOut> const& ofrAmt,
         TAmounts<TIn, TOut> const& stepAmt,
         TOut const& ownerGives) const;
+};
+
+//------------------------------------------------------------------------------
+
+// Flow is used in two different circumstances for transferring funds:
+//  o Payments, and
+//  o Offer crossing.
+// The rules for handling funds in these two cases are almost, but not
+// quite, the same.
+
+// Payment BookStep template class (not offer crossing).
+template<class TIn, class TOut>
+class BookPaymentStep
+    : public BookStep<TIn, TOut, BookPaymentStep<TIn, TOut>>
+{
+public:
+    using BookStep<TIn, TOut, BookPaymentStep<TIn, TOut>>::BookStep;
+    using BookStep<TIn, TOut, BookPaymentStep<TIn, TOut>>::qualityUpperBound;
+
+    // Never limit self cross quality on a payment.
+    bool limitSelfCrossQuality (AccountID const&, AccountID const&,
+        TOffer<TIn, TOut> const& offer, boost::optional<Quality>&,
+        FlowOfferStream<TIn, TOut>&, bool) const
+    {
+        return false;
+    }
+
+    // A payment can look at offers of any quality
+    bool checkQualityThreshold(TOffer<TIn, TOut> const& offer) const
+    {
+        return true;
+    }
+
+    // For a payment ofrInRate is always the same as trIn.
+    std::uint32_t getOfrInRate (
+        Step const*, TOffer<TIn, TOut> const&, std::uint32_t trIn) const
+    {
+        return trIn;
+    }
+
+    // For a payment ofrOutRate is always the same as trOut.
+    std::uint32_t getOfrOutRate (Step const*, TOffer<TIn, TOut> const&,
+        AccountID const&, std::uint32_t trOut) const
+    {
+        return trOut;
+    }
+
+    Quality
+    qualityUpperBound(ReadView const& v,
+        Quality const& ofrQ,
+        bool prevStepRedeems) const
+    {
+        // Charge the offer owner, not the sender
+        // Charge a fee even if the owner is the same as the issuer
+        // (the old code does not charge a fee)
+        // Calculate amount that goes to the taker and the amount charged the
+        // offer owner
+        auto rate = [&](AccountID const& id) -> std::uint32_t {
```

Comment:
Can you try the following, where we use `Rate` instead of naked `std::uint32_t`:

```
auto rate = [&](AccountID const& id) {
    if (isXRP(id) || id == this->strandDst_)
        return parityRate;
    return transferRate(v, id);
};

auto const trIn = prevStepRedeems
    ? rate(this->book_.in.account)
    : parityRate;
// Always charge the transfer fee, even if the owner is the issuer
auto const trOut = this->ownerPaysTransferFee_
    ? rate(this->book_.out.account)
    : parityRate;

Quality const q1{getRate(STAmount(trOut.value), STAmount(trIn.value))};
return composed_quality(q1, ofrQ);
```

At first, it may seem like it doesn't buy us much, but it adds type safety (and will add more once we change `getRate` to take `Rate` instances instead of two `std::uint32_t` values) and makes the code self-documenting (quick, what format is a `std::uint32_t` representing a rate and is `QUALITY_ONE` a rate or a quality?)

Note, that the `auto const iss` code around line 254 _shouldn't_ be needed, and I don't use it: simply not specifying an issuer _should_ give you the right results.